crates/bevy_asset/src/io/gated.rs

```diff
@@ -36,7 +33,7 @@ impl GateOpener {
     pub fn open<P: AsRef<Path>>(&self, path: P) {
         let mut gates = self.gates.write();
         let gates = gates
-            .entry(path.as_ref().to_path_buf())
+            .entry(Box::from(path.as_ref()))
```

Comment:
Could this be replace with an `insert`? This allocates even when there is already an existing entry.

Reply:
`insert` didn't returns the reference to the inserted value, it will need an extra `get_mut`.

Also, there is the [`raw_entry_mut`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.raw_entry_mut) mehtod that could defer the creation of the owned key, but it's unstable.
Otherwise, in `hashbrown` there is the [`entry_ref`](https://docs.rs/hashbrown/latest/hashbrown/struct.HashMap.html#method.entry_ref) method that could do that.

Reply:
yeah, `entry_ref` is exactly what I think we should be using. I wasn't aware of it. It seems to only convert from the `&T` into a `Box<T>` when inserting new value (which is exactly what we want). Aren't we using the `hashbrown` hashmaps in any case?

Well, I think this PR is already in a mergeable state, and changing from `entry` to `entry_ref` is just sugar on top.

Reply:
> Aren't we using the `hashbrown` hashmaps in any case?

`hashbrown` is present in the dependency graph of `bevy_asset` via `bevy_utils` (but I think it's about to be removed: #11721) and via `bevy_macros_utils` > `toml_edit` > `indexmap`.
So I think it's okay to use `hashbrown` here, at least until `raw_entry_mut` is stabilised. 
I'll open a PR for that after this one is merged.